R/ParamHandle.R

```diff
@@ -6,15 +6,156 @@
 #'
 #' @return [\code{\link{ParamHandle}}].
 #' @family ParamHelpers
+#' @export
+
 ParamHandle = R6Class("ParamHandle",
   inherit = ParamBase, # FIXME: Are we sure? Yes!
   public = list(
-
-    root = NULL,
+    # member variables
+    id = NULL,
+    node = NULL, # simple ParamNode the handle(pointer) point to
+    val = NULL,  # the value of the SimpleParamNode it points to. val is used for sampling
+    flatval = NULL, # if the node is itself a tree, this hold the preroot traversal of the tree
+    depend = NULL,  # depend is a list with field <id><[val][func]> which will decide if the current node is going to be activated in a sampling process, if depend$func() is True, then the sampling function is called
+    require.expr = NULL,
+    #
+    parent = NULL,
+    root = NULL,    # root has to be changed when parent changed!
+    reldepth = 0L,  # reldepth has to be updated when parent changed!
+    #
+    mand.children = NULL,
+    cond.children = NULL,
+    visitor = NULL,
 
     # constructor
-    initialize = function() {
+    initialize = function(id = NULL, node = NULL, val = NULL, depend = NULL, parent = NULL) {
+      if (is.null(id) & is.null(node)) stop("either set id or node for handle!")
+      self$id = ifelse(is.null(id), node$id, id)
+      self$node = node
+      self$val = val
+      #
+      self$depend = depend
+      self$require.expr = function(parent) {
+        if (is.null(parent$val)) return(FALSE)
+        if (is.null(self$depend)) return(TRUE)
+        return(parent$val == self$depend$val)
```

Comment:
self$parent$val is more safe.

Reply:
I think parent$val is safer since self$require.expr(parent) take a parameter parent and this parent can be something else than self$parent, see line 110 of the same file.